rllib/examples/envs/env_rendering_and_recording.py

```diff
@@ -1,135 +1,274 @@
+
+    def on_episode_end(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """Computes episode's return and compiles a video, iff best/worst in this iter.
+
+        Note that the actual logging to the EnvRunner's MetricsLogger only happens
+        at the very env of sampling (when we know, which episode was the best and
+        worst). See `on_sample_end` for the implemented logging logic.
         """
+        # Get the episode's return.
+        episode_return = episode.get_return()
+
+        # Better than the best or worse than worst Episode thus far?
+        if (
+            episode_return > self.best_episode_and_return[1]
+            or episode_return < self.worst_episode_and_return[1]
+        ):
+            # Pull all images from the temp. data of the episode.
+            images = episode.get_temporary_timestep_data("render_images")
 
-        # Just generate a random image here for demonstration purposes.
-        # Also see `gym/envs/classic_control/cartpole.py` for
-        # an example on how to use a Viewer object.
-        return np.random.randint(0, 256, size=(300, 400, 3), dtype=np.uint8)
+            # Create a video from the images by simply stacking them.
+            video = np.expand_dims(
+                np.stack(images, axis=0), axis=0
+            )  # TODO: test to make sure WandB properly logs videos.
+            # video = np.stack(images, axis=0)
 
+            # `video` is from the best episode in this cycle (iteration).
+            if episode_return > self.best_episode_and_return[1]:
+                self.best_episode_and_return = (video, episode_return)
+            # `video` is worst in this cycle (iteration).
+            else:
+                self.worst_episode_and_return = (video, episode_return)
+
+    def on_sample_end(
+        self,
+        *,
+        env_runner,
+        metrics_logger,
+        samples,
+        **kwargs,
+    ) -> None:
+        """Logs the best and worst video to this EnvRunner's MetricsLogger."""
+        # Best video.
+        metrics_logger.log_value(
+            "episode_videos_best",
+            self.best_episode_and_return[0],
+            # Do not reduce the videos (across the various parallel EnvRunners). This
+            # would not make sense (mean over the pixels?). Instead, we want to log all
+            # best videos of all EnvRunners per iteration.
+            reduce=None,
+            # B/c we do NOT reduce over the video data (mean/min/max), we need to make
+            # sure the list of videos in our MetricsLogger does not grow infinitely and
+            # gets cleared after each `reduce()` operation, meaning every time, the
+            # EnvRunner is asked to send its logged metrics.
+            clear_on_reduce=True,
+        )
+        # Worst video.
+        metrics_logger.log_value(
```

Comment:
The `log_value` at first confused me as I was expecting a single scalar value to be logged, but in comparison to the other logging methods, this makes sense, as we have only a single "variable"

Reply:
Yeah, our Tune logging API does NOT allow for specifying any data types (videos, images, etc..) this early. We have to "communicate" with it via the tensor format/shape. So there is only really `log_value`nothing else. `log_dict` and `log_n_dicts` are convenience methods to avoid having to call `log_value` a dozen times or so.